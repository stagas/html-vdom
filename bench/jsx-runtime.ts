export let jsx = () => void 0
export let jsxs = jsx
export let Fragment = null
export const setEngine = engine => {
  jsx = engine.jsx
  jsxs = engine.jsxs
  Fragment = engine.Fragment
}
