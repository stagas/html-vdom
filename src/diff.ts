import { VList } from './v'

// import { inspect } from 'util'
// const log = (...args: any[]) => console.log(...args.map(x => inspect(x, false, Infinity, true)))

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ParentLike = ParentNode | VList<any>

const $ = (P: ParentLike) => (Array.isArray(P) ? P.at(-1) : P.lastChild)

// const nextSibling = <T>(parent: ParentLike<T>, child: T | ChildNode): T | ChildNode | null =>
//   parent instanceof VList ? parent.nextSiblingOf(child as T) : (child as ChildNode).nextSibling

const _ = <T>(parent: ParentLike | T[], index: number): T | null =>
  Array.isArray(parent)
    ? (parent[index] as T) ?? null
    : ((parent as ParentNode).childNodes[index] as unknown as T) ?? null

const len = (P: ParentLike) => (Array.isArray(P) ? P.length : P.childNodes.length)

export const diff = <T>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  eq: (a: any, b: any) => boolean,
  P: ParentLike,
  b: T[],
  a: T[] = []
) => {
  // log(a, b)

  // A holds the anchor element where insertBefore(x, A) latches on
  let A: T | null

  // B holds the current element in the new list b
  let B: T | null

  // C holds a temp value of the list a that will be compared and maybe inserted
  let C: T | null

  // log(b, a)
  let i = 0

  const n = b.length
  let p = a.length

  //              p
  // a  1 2 3 4 5
  //        i
  // b  1 2 0 4 5
  //              n
  // R  1 2
  //
  // advance start pointer while equal head
  // TODO:                     set A B here
  while (i < n && i < p && eq((A = _(a, i)), (B = _(b, i)))) i++

  //          p
  // a  1 2 3 4 5
  //        i
  // b  1 2 0 4 5
  //          n
  // R  1 2
  //
  // shorten end pointer while equal tail
  // while (i < n && i < p && eq(_(a, p - 1), _(b, n - 1))) --p, --n

  // insert:
  //        A
  //        p
  // a  1 2 4 5
  //              i
  // b  1 2 0 0 0 4 5
  //              n
  // R  1 2 0 0 0
  //

  // append:
  //          A
  //          p
  // a  1 2 3
  //          i
  // b  1 2 3 4
  //            n
  //

  // prepend:
  //
  //    A
  //    p
  // a  1 2 3
  //    i
  // b  4 1 2 3
  //      n
  //

  // insert/append/prepend shortcut
  if (i == p) {
    A = _(a, p)
    for (; i < n; i++) P.insertBefore(_(b, i)!, A)
    // we are done -- rest of elements (if any) are the same
  } else {
    // replace:
    //
    //      A
    //        p
    // a  1 2 3
    //      i
    // b  1 4 3
    //        n
    //      B

    // replace + remaining tail:
    //
    //        A
    //        p
    // a  1 2 3
    //        i
    // b  1 4 5 6 3
    //            n
    // R  1 4 3
    //      B

    //      x
    //      A
    //          p
    // a  1 2 3 4
    //        i
    //        C
    //            y
    // b  1 3 2 4
    //          n
    //            B
    // R  1 3 2 4

    //      x
    //      A
    //            p
    // a  1 2 3 4
    //        C
    //        i
    //      y
    // b  1 3 4 2
    //      B
    //            n
    // R  1 3 4 2

    // moves:
    //
    //        x
    //                    p
    //        A
    // a  1 2 3 4 5 6 7 8
    //    i         y
    //    C
    // b  3 6 5 1 2 7 8 4
    //              B
    //                    n
    //
    // R  1 2 3 4 5 6 7 8
    // R  3 6 5 1 2 4 7 8
    //

    // let x = i
    // let y = i
    // for (; y < n; y++) {
    //   for (i = y; i < p; i++) {
    //     C = _(P, i)
    //     if (eq(C, B)) {
    //       P.insertBefore(C, A)
    //       break
    //     }
    //     //
    //   }

    //   while (y < n && x < p && eq((A = _(P, y)), (B = _(b, y)))) x++, y++

    //   //      A = _(P, x)
    // }

    let x = 0
    out: while (((A = _(P, i)!), (B = _(b, i)!), i++ < n)) {
      if (!eq(A, B)) {
        for (x = i; x < p; x++) {
          C = _(P, x)!
          if (eq(B, C)) {
            P.insertBefore(C, A)
            continue out
          }
        }
        for (x = i; x < n; x++) {
          C = _(b, x)
          if (eq(A, C)) {
            P.insertBefore(B, A)
            continue out
          }
        }
        if (A) P.replaceChild(B, A)
        else P.appendChild(B) //, null)
      }
    }
    p = len(P)
    i = n
    while (i++ < p) P.removeChild($(P))

    // remove tail:
    //
    //          p
    // a  1 2 3
    //      i
    // b  1 2
    //        n

    // remove head:
    //
    //      p
    // a  1 2 3
    //    i
    // b  2 3
    //    n
    // for (; i < p--; ) P.removeChild(_(P, i))
  }

  // if (P instanceof Node) {
  //   log('RESULT', P.innerHTML)
  // } else log('RESULT', [...P])
}
