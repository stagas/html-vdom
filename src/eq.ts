import { VAtom } from '.'

export const atomEq = (a: VAtom, b: VAtom) => {
  // prettier-ignore
  const eq = (
    (a != null && b != null) &&
    (
      (a === b) ||
      (
        (a[1] && b[1] && 'key' in a[1] && 'key' in b[1])
          ? a[1]!.key === b[1]!.key
          : a[0] === b[0]
      )
    )
  )
  // console.log('EQUAL', eq, a, b)
  return eq
}

export const domEq = (a: Node, b: Node) => {
  return a === b
}
