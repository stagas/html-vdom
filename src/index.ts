export * from './h'
export * from './render'
