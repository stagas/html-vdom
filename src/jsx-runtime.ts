export * from 'html-jsx'
import type * as jsxi from 'html-jsx'
import { createProps, updateProps } from './props'
export { createProps, updateProps }

declare module 'html-jsx' {
  // eslint-disable-next-line
  interface DOMAttributes<T> extends JSX.IntrinsicAttributes {}
}

/**
 * JSX types
 * @private
 */
declare global {
  namespace JSX {
    /**
     * The type returned by our `jsx` factory.
     * @private
     */
    type Element = VKid

    interface IntrinsicElements extends jsxi.IntrinsicElements {
      /**
       * This allows for any tag to be used.
       * @private
       */
      [k: string]: unknown
    }

    interface IntrinsicAttributes {
      /**
       * Pass an object that will be assigned the DOM element reference in its `current` property.
       * @private
       */
      ref?: VRef<unknown> | null

      /**
       * List index key - each item's `key` must be unique.
       * @private
       */
      key?: string | number

      /**
       * Children.
       * @private
       */
      children?: JSX.Element

      /**
       * Sets the `innerHTML` of an element to the **exact** string **without** escaping.
       */
      innerHTML?: string
    }

    interface HTMLAttributes<T> extends jsxi.HTMLAttributes<T> {}
    interface SVGAttributes<T> extends jsxi.SVGAttributes<T> {}
    interface DOMAttributes<T> extends jsxi.DOMAttributes<T> {}
  }
}

export type $<T> = (props: T & { children?: any }) => JSX.Element

export interface VRef<T> {
  current: T
}

type Fn = (...args: any) => any
type Doc = (tag: string, opts?: ElementCreationOptions) => Element
export type VFn = (props: any) => VKid
type VKid = VKids | VNode<any> | string | number | boolean | null | undefined | void
type VKids =
  & VKid[]
  & Partial<{
    running: boolean
    dom: El[]
    flatDom: DomEl[]
    keyed: Map<string, number>
    mapped: Map<El, VKid>
  }>
type DomEl = Element | CharacterData | ChildNode
type El = DomEl | Chunk
type TargetEl = El | ShadowRoot
type VAny = VNode<any>
export type Hook = Fn & { fn: Fn; onremove?: Fn } & Record<string, any>
export type Props = Record<string, any>
type VNode<T extends string | symbol | typeof Text | typeof Comment | VFn> = {
  kind: T
  props: Props
  key?: string
  hook?: Hook
  keep?: boolean
}

const anchor = new Comment()
export const Fragment = Symbol()
export const jsx = (kind: any, props: any, key: any) =>
  kind === Fragment
    ? props.children
    : { kind, props, key } as VNode<typeof kind>
export const jsxs = jsx

export let hook: Hook
const createHook = () =>
  function current(fn: Fn = (current as Hook).fn) {
    const prev = hook
    hook = current as Hook
    hook.fn = fn
    fn()
    hook = prev
  } as Hook

class Chunk extends Array {
  firstChild?: any
  dom: El[] = []
  get last() {
    return this.dom.at(-1)
  }
  get nextSibling(): El | null | undefined {
    return this.last?.nextSibling
  }
  insertBefore(x: any) {
    this.unshift(x)
  }
  appendChild(x: any) {
    this.push(x)
  }
  after(x: Node) {
    this.last?.after(x)
  }
  save() {
    this.dom = [...this]
  }
  remove() {
    this.dom.forEach(el => el.remove())
    this.splice(0)
  }
  removeChild(x: any) {
    let i = this.indexOf(x)
    ~i && this.splice(i, 1)
    i = this.dom.indexOf(x)
    ~i && this.dom.splice(i, 1)
  }
}

export type { Chunk, VNode }

const { TEXT_NODE, COMMENT_NODE } = document
const html = document.createElement.bind(document)
const svg = document.createElementNS.bind(document, 'http://www.w3.org/2000/svg')
const forceArray = <T>(x: any, withNull: boolean): T =>
  (Array.isArray(x) ? x : x == null && !withNull ? [] : [x]) as unknown as T
const flatDom = (arr: El[], res: DomEl[] = []) => {
  for (const el of arr) {
    if ((el as Chunk).dom) res.push(...flatDom((el as Chunk).dom!))
    else res.push(el as DomEl)
  }
  return res
}

const prevs = new WeakMap()
export const render = (n: VKid, el: TargetEl, doc: Doc = html, withNull = false) =>
  reconcile(el, forceArray(n, withNull), prevs.get(el), doc)

const reconcile = (parent: TargetEl, nk: VKids, pk: VKids | VKid, doc: Doc) => {
  if ((pk as VKids)?.running) {
    console.warn('attempt to reconcile lane which is reconciling')
    return
  }

  if (pk === nk) nk = [...nk]

  prevs.set(parent, nk)

  nk.running = true
  nk.dom = Array(nk.length)
  nk.keyed = new Map()
  nk.mapped = new Map()

  if (Array.isArray(pk)) {
    const keep = new Set()
    for (let i = 0, n, el, p, pel, k, pi; i < nk.length; i++) {
      n = nk[pi = i]
      k = (n as VAny)?.key
      if (k != null) {
        nk.keyed.set(k, i)
        pi = pk.keyed!.get(k) ?? -1
      }
      p = pk[pi]
      pel = pk.dom![pi]
      nk.dom[i] = el = create(doc, n, p, pel)
      if (el === pel) keep.add(pel)
    }
    for (const pel of pk.dom!) {
      if (!keep.has(pel)) {
        if (parent instanceof Chunk) parent.removeChild(pel)
        pel.remove()
        ;(pk.mapped!.get(pel) as VNode<VFn>)?.hook?.onremove?.()
      }
    }
  } else {
    for (let i = 0, n, el, k; i < nk.length; i++) {
      n = nk[i]
      k = (n as VAny)?.key
      if (k != null) nk.keyed.set(k, i)
      nk.dom[i] = el = create(doc, n)
      nk.mapped.set(el, n)
    }
  }

  nk.flatDom = flatDom(nk.dom)
  if ((pk as VKids)?.flatDom)
    diff(parent, nk.flatDom, (pk as VKids)!.flatDom!)
  else
    nk.flatDom.forEach(el => parent.appendChild(el))

  nk.running = false
}

const diff = (parent: TargetEl, n: DomEl[], p: DomEl[], i = 0, len = n.length, el?: DomEl, last?: DomEl) => {
  for (; i < len; i++) {
    el = n[i]
    if (p[i] === el) last = el
    else if (!i) parent.insertBefore(last = el, parent.firstChild)
    else last!.after(last = el)
  }
}

const create = (doc: Doc, n: VKid, p?: VKid, pel?: El | null) => {
  let el: El | null
  switch (typeof n) {
    case 'string':
    case 'number':
      if ((pel as Node)?.nodeType === TEXT_NODE) {
        if (p != n) (pel as Node).nodeValue = n as string
        return pel as El
      }
      el = new Text(n as string)
      return el
    case 'object':
      if (!n) break
      if (Array.isArray(n)) {
        if (pel && Array.isArray(p)) el = pel as Chunk
        else el = new Chunk()
        reconcile(el, n, p, doc)
        el.save()
      } else if (typeof n.kind === 'string') {
        // maybe switch to svg namespace
        if (
          ( // if we pass explicit element reference, and is the right tag, use that
            n.props.ref?.current
            && n.props.ref.current.tagName.toLowerCase() === n.kind
            && (el = n.props.ref.current as El)
          ) || (
            // if the previous element is of the same kind, use that
            pel
            && (p as VNode<string>).kind === n.kind
            && (el = pel)
          )
        ) {
          // only update props if we have the element
          updateProps(el as Element, n.kind, n.props)
        } else {
          if (n.kind === 'svg') doc = svg
          // otherwise create an element and props
          el = doc(n.kind, 'is' in n.props ? { is: n.props.is } : void 0)
          // maybe switch to xhtml namespace
          if (n.kind === 'foreignObject') doc = html
          createProps(el as Element, n.kind, n.props)
        }
        // render children
        render(n.props.children, el, doc)
        // add ref
        if ('ref' in n.props && el !== n.props.ref.current) n.props.ref.current = el
      } else {
        let initial = true
        if (!((el = pel!) && (n.hook = (p as VNode<VFn>).hook))) {
          el = new Chunk()
          n.hook = createHook()
        }
        n.hook(() => {
          let next!: ChildNode | null
          // next = el!.nextSibling as ChildNode
          // initial = true
          // const parent = (el as any)!.parentNode
          if (!initial && !(next = el!.nextSibling as ChildNode)) el!.after(next = anchor)
          // el?.remove()
          // ;(el as Chunk)!.dom!.splice(0)
          // ;(el as Chunk).splice(0)
          // }
          // console.log('P', el)
          render(n.kind(n.props), el!, doc, true)
          ;(el as Chunk).save()
          // console.log(el)
          if (!initial && next) {
            for (const e of flatDom(el as Chunk)) next!.before(e)
            next === anchor && next.remove()
          } else {
            initial = false
          }
        })
      }
      return el
  }
  if ((pel as Node)?.nodeType === COMMENT_NODE) return pel as El
  el = new Comment()
  return el
}
