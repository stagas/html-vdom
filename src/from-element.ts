import * as jsxi from 'html-jsx'
import { jsx } from './jsx-runtime'

export type Constructor<T> = new(...args: any[]) => T

const Registry: WeakMap<typeof HTMLElement, any> = new WeakMap()

export interface Options<T> extends ElementDefinitionOptions {
  interface?: Constructor<T>
}

export type Component<T, I> = (props: Partial<Omit<T & I, keyof HTMLElement>> & jsxi.HTMLAttributes<T>) => any

export const fromElement = <T extends HTMLElement, I = HTMLElement>(
  Element: Constructor<T>,
  options: Options<I> = {},
): Component<T, I> => {
  let fn: Component<T, I> = Registry.get(Element)
  if (fn) return fn

  const tag = 'x-' + Element.name.toLowerCase() + ((Math.random() * 1e7) | 0).toString(36)
  const ctor = class extends Element {}
  customElements.define(tag, ctor, options)
  if (options.extends)
    fn = props => jsx(options.extends, Object.assign(props, { is: tag }), void 0)
  else
    fn = props => jsx(tag, props, void 0)
  Registry.set(Element, fn)
  ;(fn as any).toString = () => tag

  return fn
}
