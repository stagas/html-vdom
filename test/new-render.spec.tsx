// @env browser
/** @jsxImportSource ../src */
import { Hook, hook, render } from '../src/jsx-runtime'

const r = (jsx: any, div = document.createElement('div')) => {
  render(jsx, div)
  return div
}

const o = (jsx: any, div?: any) => r(jsx, div).innerHTML

describe('all', () => {
  describe('render', () => {
    it('simple', () => {
      expect(o(<div />)).toMatchSnapshot()
    })

    it('functional simple', () => {
      const Foo = () => <div></div>
      expect(o(<Foo />)).toMatchSnapshot()
    })

    it('functional nested', () => {
      const Bar = ({ label }: any) => <p>{label}</p>
      const Foo = () => (
        <div>
          <Bar label="a" />
          <Bar label="b" />
        </div>
      )
      expect(o(<Foo />)).toMatchSnapshot()
    })

    it('two levels', () => {
      expect(
        o(
          <div>
            <p>hi</p>
          </div>
        )
      ).toMatchSnapshot()
    })

    it('fragment', () => {
      expect(
        o(
          <>
            <a />
            <b />
          </>
        )
      ).toMatchSnapshot()
    })

    it('text child', () => {
      expect(o(<p>hi</p>)).toMatchSnapshot()
    })

    it('multiple text childs', () => {
      const v = 'there'
      expect(o(<p>hi {v}</p>)).toMatchSnapshot()
    })
  })

  describe('patch', () => {
    it('replace node', () => {
      const el = r(<div />)
      expect(el.innerHTML).toMatchSnapshot()
      expect(o(<p />, el)).toMatchSnapshot()
    })

    it('functional update', () => {
      let i = 0
      const Foo = () => <div>{i++}</div>
      const el = r(<Foo />)
      const first = el.firstChild
      expect(el.innerHTML).toMatchSnapshot()
      expect(o(<Foo />, el)).toMatchSnapshot()
      expect(el.firstChild === first).toBe(true)
    })

    it('hook update', () => {
      let i = 0
      let h!: Hook
      const Foo = () => {
        h = hook
        return <div>{i++}</div>
      }
      const el = r(<Foo />)
      const first = el.firstChild
      expect(el.innerHTML).toMatchSnapshot()
      h()
      expect(el.innerHTML).toMatchSnapshot()
      expect(el.firstChild === first).toBe(true)
    })

    it('hook onremove', () => {
      let i = 0
      let h!: Hook
      const Foo = () => {
        h = hook
        return <div>{i++}</div>
      }
      const el = r(<Foo />)
      expect(el.innerHTML).toMatchSnapshot()
      let count = 0
      h.onremove = () => count++
      // console.log('---------------')
      h()
      expect(el.innerHTML).toMatchSnapshot()
      expect(count).toBe(0)
      expect(o(<></>, el)).toMatchSnapshot()
      expect(count).toBe(1)
    })

    it('hook is same object', () => {
      let h!: Hook
      const Foo = () => {
        h = hook
        h.i ??= 0
        return <div>{h.i++}</div>
      }
      const el = r(<Foo />)
      expect(el.innerHTML).toMatchSnapshot()
      h()
      expect(el.innerHTML).toMatchSnapshot()
      h()
      expect(el.innerHTML).toMatchSnapshot()
    })

    it('hook update single first undefined', () => {
      let i = 0
      let h!: Hook
      const Foo = () => {
        h = hook
        return i++ < 1 ? void 0 : i
      }
      const el = r(<Foo />)
      expect(el.innerHTML).toMatchSnapshot()
      h()
      expect(el.innerHTML).toMatchSnapshot()
    })

    it('hook update sibling first undefined', () => {
      let i = 0
      let h!: Hook
      const Foo = () => {
        if (!h) h = hook
        return i++ < 1 ? void 0 : i
      }
      const el = r(
        <>
          <Foo />
          <Foo />
        </>
      )
      expect(el.innerHTML).toMatchSnapshot()
      h()
      expect(el.innerHTML).toMatchSnapshot()
    })

    it('hook siblings update', () => {
      let i = 0
      let h!: Hook
      const Foo = () => {
        if (!h) h = hook
        return <div>{i++}</div>
      }
      const el = r(
        <>
          <Foo />
          <Foo />
        </>
      )
      const first = el.firstChild
      expect(el.innerHTML).toMatchSnapshot()
      h()
      expect(el.innerHTML).toMatchSnapshot()
      expect(el.firstChild === first).toBe(true)
    })

    it('hook siblings keyed update', () => {
      let i = 0
      let h!: Hook
      const Foo = () => {
        if (!h) h = hook
        return <div>{i++}</div>
      }
      const el = r(
        <>
          <Foo key="a" />
          <Foo key="b" />
        </>
      )
      const first = el.firstChild
      expect(el.innerHTML).toMatchSnapshot()
      h()
      expect(el.innerHTML).toMatchSnapshot()
      expect(el.firstChild === first).toBe(true)
    })

    it('functional nested update', () => {
      let i = 0
      const Bar = ({ label }: any) => <p>{label}</p>
      const Foo = () => (
        <div>
          <Bar label={++i} />
          <Bar label={++i} />
        </div>
      )
      const el = r(<Foo />)
      const first = el.firstChild!
      const firstFirst = first.firstChild
      expect(el.innerHTML).toMatchSnapshot()
      expect(o(<Foo />, el)).toMatchSnapshot()
      expect(el.firstChild === first).toBe(true)
      expect(el.firstChild!.firstChild === firstFirst).toBe(true)
    })

    it('functional nested drop', () => {
      let i = 0
      const Bar = ({ label }: any) => <p>{label}</p>
      const Foo = ({ count }: any) => (
        <div>
          {Array(count)
            .fill(0)
            .map(() => <Bar label={++i} />)}
        </div>
      )
      const el = r(<Foo count={2} />)
      const first = el.firstChild!
      const firstFirst = first.firstChild
      expect(el.innerHTML).toMatchSnapshot()
      expect(o(<Foo count={1} />, el)).toMatchSnapshot()
      expect(el.firstChild === first).toBe(true)
      expect(el.firstChild!.firstChild === firstFirst).toBe(true)
    })

    it('drop one node', () => {
      const el = r(
        <>
          <a />
          <b />
          <c />
        </>
      )
      const els = [...el.childNodes]
      expect(el.innerHTML).toMatchSnapshot()
      expect(
        o(
          <>
            <a />
            <b />
          </>,
          el
        )
      ).toMatchSnapshot()
      const res = [...el.childNodes]
      expect(els[0]).toBe(res[0])
      expect(els[1]).toBe(res[1])
    })

    it('two drop one', () => {
      const el = r(
        <>
          <a />
          <b />
        </>
      )
      const els = [...el.childNodes]
      expect(el.innerHTML).toMatchSnapshot()
      expect(
        o(
          <>
            <a />
          </>,
          el
        )
      ).toMatchSnapshot()
      const res = [...el.childNodes]
      expect(els[0]).toBe(res[0])
    })

    it('two drop first', () => {
      const el = r(
        <>
          <a />
          <b />
        </>
      )
      // const els = [...el.childNodes]
      expect(el.innerHTML).toMatchSnapshot()
      expect(
        o(
          <>
            <b />
          </>,
          el
        )
      ).toMatchSnapshot()
      // const res = [...el.childNodes]
      // expect(els[0]).toBe(res[0])
    })

    it('keyed two drop first', () => {
      const el = r(
        <>
          <a key="a" />
          <b key="b" />
        </>
      )
      const els = [...el.childNodes]
      expect(el.innerHTML).toMatchSnapshot()
      expect(
        o(
          <>
            <b key="b" />
          </>,
          el
        )
      ).toMatchSnapshot()
      const res = [...el.childNodes]
      expect(els[1]).toBe(res[0])
    })

    it('replace text', () => {
      const el = r(<p>hi</p>)
      const first = el.firstChild!
      const text = first.firstChild
      expect(el.innerHTML).toMatchSnapshot()
      expect(o(<p>there</p>, el)).toMatchSnapshot()
      expect(el.firstChild).toBe(first)
      expect(el.firstChild!.firstChild).toBe(text)
    })

    it('insert text node', () => {
      const el = r(<p>hi</p>)
      const first = el.firstChild!
      // const text = first.firstChild
      const v = 'hey'
      expect(el.innerHTML).toMatchSnapshot()
      expect(o(<p>{v} hi</p>, el)).toMatchSnapshot()
      expect(el.firstChild).toBe(first)
      // expect(el.firstChild!.firstChild).toBe(text)
    })

    it('move keyed node', () => {
      const el = r(
        <>
          <li key="a" />
          <li key="b" />
        </>
      )
      const lis = [...el.querySelectorAll('li')]
      expect(el.innerHTML).toMatchSnapshot()
      expect(
        o(
          <>
            <li key="b" />
            <li key="a" />
          </>,
          el
        )
      ).toMatchSnapshot()
      const res = [...el.querySelectorAll('li')]
      expect(lis[0]).toBe(res[1])
      expect(lis[1]).toBe(res[0])
    })

    it('move key ref node', () => {
      const a = { a: true }
      const b = { b: true }

      const el = r(
        <>
          <li key={a} />
          <li key={b} />
        </>
      )
      const lis = [...el.querySelectorAll('li')]
      expect(el.innerHTML).toMatchSnapshot()
      expect(
        o(
          <>
            <li key={b} />
            <li key={a} />
          </>,
          el
        )
      ).toMatchSnapshot()
      const res = [...el.querySelectorAll('li')]
      expect(lis[0]).toBe(res[1])
      expect(lis[1]).toBe(res[0])
    })

    it('insert keyed node', () => {
      const el = r(
        <>
          <li key="a" />
          <li key="b" />
        </>
      )
      const lis = [...el.querySelectorAll('li')]
      expect(el.innerHTML).toMatchSnapshot()
      expect(
        o(
          <>
            <li key="c" />
            <li key="a" />
          </>,
          el
        )
      ).toMatchSnapshot()
      const res = [...el.querySelectorAll('li')]
      expect(lis[0]).toBe(res[1])
      expect(lis[1]).not.toBe(res[0])
    })

    it('insert keyed node svg', () => {
      const el = r(
        <svg>
          <rect key="a" />
          <rect key="b" />
        </svg>
      )
      const lis = [...el.querySelectorAll('rect')]
      expect(el.innerHTML).toMatchSnapshot()
      expect(
        o(
          <svg>
            <rect key="c" />
            <rect key="a" />
          </svg>,
          el
        )
      ).toMatchSnapshot()
      const res = [...el.querySelectorAll('rect')]
      expect(lis[0]).toBe(res[1])
      expect(lis[1]).not.toBe(res[0])
    })

    it('insert replace keyed node', () => {
      const el = r(
        <>
          <li key="a" />
          <li key="b" />
        </>
      )
      const lis = [...el.querySelectorAll('li')]
      expect(el.innerHTML).toMatchSnapshot()
      expect(
        o(
          <>
            <li key="d" />
            <li key="a" />
            <li key="c" />
          </>,
          el
        )
      ).toMatchSnapshot()
      const res = [...el.querySelectorAll('li')]
      expect(lis[0]).toBe(res[1])
    })

    it('replace keyed node', () => {
      const el = r(
        <>
          <li key="a" />
          <li key="b" />
        </>
      )
      const lis = [...el.querySelectorAll('li')]
      expect(el.innerHTML).toMatchSnapshot()
      expect(
        o(
          <>
            <li key="a" />
            <li key="c" />
          </>,
          el
        )
      ).toMatchSnapshot()
      const res = [...el.querySelectorAll('li')]
      expect(lis[0]).toBe(res[0])
      expect(lis[1]).not.toBe(res[1])
    })

    it('replace multiple keyed node', () => {
      const el = r(
        <>
          <li key="a" />
          <li key="b" />
        </>
      )
      const lis = [...el.querySelectorAll('li')]
      expect(el.innerHTML).toMatchSnapshot()
      {
        expect(
          o(
            <>
              <li key="a" />
              <li key="c" />
            </>,
            el
          )
        ).toMatchSnapshot()
        const res = [...el.querySelectorAll('li')]
        expect(lis[0]).toBe(res[0])
        expect(lis[1]).not.toBe(res[1])
      }
      {
        expect(
          o(
            <>
              <li key="b" />
              <li key="c" />
            </>,
            el
          )
        ).toMatchSnapshot()
        const res = [...el.querySelectorAll('li')]
        expect(lis[1]).not.toBe(res[0])
        expect(lis[0]).not.toBe(res[1])
      }
    })

    it('deep move keyed node', () => {
      const el = r(
        <div>
          <li key="a" />
          <li key="b" />
        </div>
      )
      const first = el.firstChild
      const lis = [...el.querySelectorAll('li')]
      expect(el.innerHTML).toMatchSnapshot()
      expect(
        o(
          <div>
            <li key="b" />
            <li key="a" />
          </div>,
          el
        )
      ).toMatchSnapshot()
      const res = [...el.querySelectorAll('li')]
      expect(el.firstChild).toBe(first)
      expect(lis[0]).toBe(res[1])
      expect(lis[1]).toBe(res[0])
    })

    it('move outer keyed node', () => {
      const el = r(
        <>
          <div key="a">
            <li key="a" />
            <li key="b" />
          </div>
          <div key="b">
            <li key="a" />
            <li key="b" />
          </div>
        </>
      )
      const divs = [...el.querySelectorAll('div')]
      expect(el.innerHTML).toMatchSnapshot()
      expect(
        o(
          <>
            <div key="b">
              <li key="a" />
              <li key="b" />
            </div>
            <div key="a">
              <li key="a" />
              <li key="b" />
            </div>
          </>,
          el
        )
      ).toMatchSnapshot()
      const res = [...el.querySelectorAll('div')]
      expect(divs[0]).toBe(res[1])
      expect(divs[1]).toBe(res[0])
    })

    it('replace inner node', () => {
      const el = r(
        <div>
          <p></p>
        </div>
      )
      const first = el.firstChild
      expect(el.innerHTML).toMatchSnapshot()
      expect(
        o(
          <div>
            <a></a>
          </div>,
          el
        )
      ).toMatchSnapshot()
      expect(el.firstChild).toBe(first)
    })
  })
})
