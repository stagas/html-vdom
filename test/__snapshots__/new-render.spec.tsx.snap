// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all patch deep move keyed node 1`] = `"<div><li></li><li></li></div>"`;

exports[`all patch deep move keyed node 2`] = `"<div><li></li><li></li></div>"`;

exports[`all patch drop one node 1`] = `"<a></a><b></b><c></c>"`;

exports[`all patch drop one node 2`] = `"<a></a><b></b>"`;

exports[`all patch functional nested drop 1`] = `"<div><p>1</p><p>2</p></div>"`;

exports[`all patch functional nested drop 2`] = `"<div><p>3</p></div>"`;

exports[`all patch functional nested update 1`] = `"<div><p>1</p><p>2</p></div>"`;

exports[`all patch functional nested update 2`] = `"<div><p>3</p><p>4</p></div>"`;

exports[`all patch functional update 1`] = `"<div>0</div>"`;

exports[`all patch functional update 2`] = `"<div>1</div>"`;

exports[`all patch hook is same object 1`] = `"<div>0</div>"`;

exports[`all patch hook is same object 2`] = `"<div>1</div>"`;

exports[`all patch hook is same object 3`] = `"<div>2</div>"`;

exports[`all patch hook onremove 1`] = `"<div>0</div>"`;

exports[`all patch hook onremove 2`] = `"<div>1</div>"`;

exports[`all patch hook onremove 3`] = `""`;

exports[`all patch hook siblings keyed update 1`] = `"<div>0</div><div>1</div>"`;

exports[`all patch hook siblings keyed update 2`] = `"<div>2</div><div>1</div>"`;

exports[`all patch hook siblings update 1`] = `"<div>0</div><div>1</div>"`;

exports[`all patch hook siblings update 2`] = `"<div>2</div><div>1</div>"`;

exports[`all patch hook update 1`] = `"<div>0</div>"`;

exports[`all patch hook update 2`] = `"<div>1</div>"`;

exports[`all patch hook update sibling first undefined 1`] = `"<!---->2"`;

exports[`all patch hook update sibling first undefined 2`] = `"32"`;

exports[`all patch hook update single first undefined 1`] = `"<!---->"`;

exports[`all patch hook update single first undefined 2`] = `"2"`;

exports[`all patch insert text node 1`] = `"<p>hi</p>"`;

exports[`all patch insert text node 2`] = `"<p>hey hi</p>"`;

exports[`all patch keyed two drop first 1`] = `"<a></a><b></b>"`;

exports[`all patch keyed two drop first 2`] = `"<b></b>"`;

exports[`all patch move keyed node 1`] = `"<li></li><li></li>"`;

exports[`all patch move keyed node 2`] = `"<li></li><li></li>"`;

exports[`all patch move outer keyed node 1`] = `"<div><li></li><li></li></div><div><li></li><li></li></div>"`;

exports[`all patch move outer keyed node 2`] = `"<div><li></li><li></li></div><div><li></li><li></li></div>"`;

exports[`all patch replace inner node 1`] = `"<div><p></p></div>"`;

exports[`all patch replace inner node 2`] = `"<div><a></a></div>"`;

exports[`all patch replace node 1`] = `"<div></div>"`;

exports[`all patch replace node 2`] = `"<p></p>"`;

exports[`all patch replace text 1`] = `"<p>hi</p>"`;

exports[`all patch replace text 2`] = `"<p>there</p>"`;

exports[`all patch two drop first 1`] = `"<a></a><b></b>"`;

exports[`all patch two drop first 2`] = `"<b></b>"`;

exports[`all patch two drop one 1`] = `"<a></a><b></b>"`;

exports[`all patch two drop one 2`] = `"<a></a>"`;

exports[`all render fragment 1`] = `"<a></a><b></b>"`;

exports[`all render functional nested 1`] = `"<div><p>a</p><p>b</p></div>"`;

exports[`all render functional simple 1`] = `"<div></div>"`;

exports[`all render multiple text childs 1`] = `"<p>hi there</p>"`;

exports[`all render simple 1`] = `"<div></div>"`;

exports[`all render text child 1`] = `"<p>hi</p>"`;

exports[`all render two levels 1`] = `"<div><p>hi</p></div>"`;
